src/modules/navigator/precland.cpp

```diff
@@ -522,11 +522,13 @@ bool PrecLand::check_state_conditions(PrecLandState state)
 		} else {
 			// if not already in this state, need to be above target to enter it
 			return _target_pose_valid && _target_pose.abs_pos_valid
-			       && fabsf(_target_pose.x_rel) < _param_hacc_rad.get() && fabsf(_target_pose.y_rel) < _param_hacc_rad.get();
+			       && fabsf(_target_pose.x_abs - vehicle_local_position->x) < _param_hacc_rad.get()
+			       && fabsf(_target_pose.y_abs - vehicle_local_position->y) < _param_hacc_rad.get();
```

Comment:
The use of `_target_pose.*_rel` was deliberate here. By using relative positions for the check you can be sure that you are actually over the target. If you use absolute data you can have the situation where e.g. you measure the target once, with some offset from the vehicle. The vehicle flies to that location but for some reason does not detect the target again, and thus precland still has the same old target pose message. The absolute x/y now matches the vehicle's and the check passes, while the relative one would tell you that something is not right.

Reply:
I don't understand. I consider `*_rel` and `*_abs` fields differ only by the coordinate frame, don't they? So they are the same data (the "real" detected position of the target).
The "up-to-dateness" of the message is checked by the flag `_target_pose_valid`.

Reply:
`_target_pose_valid` is marked `false` after `PLD_BTOUT` seconds.

It's true that `*_rel` and `*_abs` are just in different coordinate frames. What's important is when this transformation is done. It is `landing_target_estimator` that fills the field if and only if a new measurement is received and fused. If, for some time period (< `PLD_BTOUT`), `landing_target_estimator` does not fuse a new measurement, the content of `landing_target_pose.msg` stays the same. If you look at the absolute values in this message you may come to the conclusion that you are directly above the target. If you look at the relative values you will realize that you are not, due to the fact that they do not say ~0, but they would be if you were on top of the beacon and getting measurements.

Reply:
`LandingTargetEstimator.cpp` doesn't publish new `landing_target_pose` message, if there is no new data (line 122):
```c++
	if (!_new_irlockReport) {
		// nothing to do
		return;
	}
```

Also, the only condition to publish `*_rel` fields is if we have a valid local position (line 217):
```
			if (_vehicleLocalPosition_valid && _vehicleLocalPosition.xy_valid) {
				_target_pose.x_abs = x + _vehicleLocalPosition.x;
				_target_pose.y_abs = y + _vehicleLocalPosition.y;
				_target_pose.z_abs = dist + _vehicleLocalPosition.z;
				_target_pose.abs_pos_valid = true;

			} else {
				_target_pose.abs_pos_valid = false;
			}
```

The calculation method, however, is the same as in the code above.

So, in my opinion, the correct way to determine if we are above the target right now is not using `*_rel` instead of `*_abs` (which are actually the same data), but to make sure, that we have a "fresh" `landing_target_pose` message:

```cpp
bool updated = false;
orb_check(_targetPoseSub, &updated);
_target_pose_updated = updated;

// ...

return _target_pose_valid && _target_pose_updated
```

Reply:
I don't see the benefit of using `(x_rel + _vehicleLocalPosition.x) - _vehicleLocalPosition.x` with an additional applicability check instead of just `x_rel`.

What you care about is the relative position of the beacon. This is also the quantity that is being estimated by `landing_target_estimator`. The abs pos is a derived value that is not relevant in this computation.

Reply:
OK, the only thing I wanted is to make it work with the `LANDING_TARGET` message, which doesn't have "relative x, y" fields.
We can recalculate the relative values after receiving the `LANDING_TAGET` message, but where? Is having local position subscriber in `mavlink_receiver` module and recalculating it there a good idea?

Reply:
I see. How is the absolute position of the target expected to be computed by the companion computer? It seems like it needs to get the local position from PX4 to add that to the relative target position (its observation). That doesn't seem ideal either.

This makes me realize that you need to check the `frame` field of the mavlink message. What you're doing now is only true for `MAV_FRAME_LOCAL_NED`. If its in `MAV_FRAME_BODY_NED` you need to fill in the `rel` fields. And in global you have to first transform it to local.

Reply:
But the companion computer MUST use the local position or at least attitude for the following reason: it needs to compensate vehicle angles to get the correct landing target position.

Also, even when using `MAV_FRAME_BODY_NED` we also have calculate the absolute position in `mavlink_receiver`, because currently `precland.cpp` requires the both relative AND absolute positions to work (that is what I was trying to fix).

So what do you think, where such a transformation can be performed? For me, `mavlink_receiver` doesn't seem the best place for it (even `precland.cpp` is better).

Reply:
> But the companion computer MUST use the local position or at least attitude for the following reason

Valid point.

> For me, mavlink_receiver doesn't seem the best place for it

Agreed. But I don't like it in `precland` either. That's a control module that should not have to worry about where the data came from and should need to do as little transformations as possible.

Perhaps it would be better if `landing_target_estimator` is the consumer of the mavlink message and constructs a `landing_target_pose` message with all the required data. Consider also the case where the mavlink message only contains angles but no valid position.

Reply:
So, you mean we have to add a new message, like `landing_target.msg`, and publish it from `mavlink_receiver` to `landing_target_estimator` so it just calculates relative or absolute positions?

Doesn't it seem too complicated?

Reply:
If we only want to support `MAV_FRAME_BODY_NED` I see how it might look like overkill.

If we want to support the other frames as well, and `LANDING_TARGET` messages that only contain angles, it is not. Now you're doing conversions and estimating things that belong in neither `mavlink_receiver` nor `precland`.

Reply:
@ndepal 
> That's a control module that should not have to worry about where the data came from and should need to do as little transformations as possible.

I examined another modules, in particular, `mc_pos_control`, and though it's a control module, it gets velocity setpoint + velocity frame and does the transformations itself.

What do you think about adding `position_frame` (and `velocity_frame`) fields to `landing_target_pose`, that determines position's coordinate frame, so `precland` could calculate all needed data?

>It's true that *_rel and *_abs are just in different coordinate frames. What's important is when this transformation is done.

We can check the time of the transformation by the `timestamp` field of the `landing_target_pose`.